pkg/datastore/clickhouse.go

```diff
@@ -0,0 +1,192 @@
+		Addr:            []string{fmt.Sprintf("%s:%s", config.Host, config.Port)},
+		Auth:            clickhouse.Auth{Database: config.Database, Username: config.Username, Password: config.Password},
+		MaxOpenConns:    config.MaxOpenConn,
+		MaxIdleConns:    config.MaxIdleConn,
+		ConnMaxLifetime: time.Duration(config.MaxConnLife),
+	})
+	if err != nil {
+		return ClickHouseDB{}, err
+	}
+
+	if err := connect.Ping(context.Background()); err != nil {
+		if exception, ok := err.(*clickhouse.Exception); ok {
+			logger.Errorf("[%d] %s \n%s\n", exception.Code, exception.Message, exception.StackTrace)
+		}
+
+		return ClickHouseDB{}, err
+	}
+
+	go pushClickhouseConnMetrics(config.Database, config.Host, connect)
+
+	db := ClickHouseDB{Conn: connect, config: config}
+
+	return db, nil
+}
+
+// pushConnMetrics pushes CLICKHOUSE connection pool values to metrics for every 100 millisecond
+func pushClickhouseConnMetrics(database, hostname string, db driver.Conn) {
+	for {
+		stats := db.Stats()
+		clickhouseOpen.WithLabelValues(database, hostname).Set(float64(stats.Open))
+		clickhouseIdle.WithLabelValues(database, hostname).Set(float64(stats.Idle))
+		time.Sleep(pushMetricDuration * time.Millisecond)
+	}
+}
+
+// HealthCheck pings the clickHouse instance in gorm. If the ping does not return an error, the healthCheck status will be set to UP,
+// else the healthCheck status will be DOWN
+func (c ClickHouseDB) HealthCheck() types.Health {
+	resp := types.Health{
+		Name:   ClickHouse,
+		Status: pkg.StatusDown,
+		Host:   c.config.Host,
+	}
+	// The following check is for the condition when the connection to CLICKHOUSE has not been made during initialization
+	if c.Conn == nil {
+		return resp
+	}
+
+	err := c.Conn.Ping(context.Background())
+	if err != nil {
+		c.logger.Error(errors.HealthCheckFailed{Dependency: ClickHouse, Err: err})
+		return resp
+	}
+
+	resp.Status = pkg.StatusUp
+	resp.Details = c.Conn.Stats()
+
+	return resp
+}
+
+// Query executes a query that returns rows, typically a SELECT.
+// The args are for any placeholder parameters in the query.
+func (c *ClickHouseDB) Query(ctx context.Context, query string, args ...interface{}) (driver.Rows, error) {
+	if c == nil || c.Conn == nil {
+		return nil, errors.CLICKHOUSENotInitialized
+	}
+
+	begin := time.Now()
+	rows, err := c.Conn.Query(ctx, query, args...)
+
+	c.monitorQuery(begin, query)
+
+	return rows, err
+}
+
+// Exec executes a query without returning any rows.
+// The args are for any placeholder parameters in the query.
+func (c *ClickHouseDB) Exec(ctx context.Context, query string, args ...interface{}) error {
+	if c == nil || c.Conn == nil {
+		return errors.CLICKHOUSENotInitialized
+	}
+
+	begin := time.Now()
+
+	err := c.Conn.Exec(ctx, query, args...)
+
+	c.monitorQuery(begin, query)
+
+	return err
+}
+
+// QueryRow executes a query that is expected to return at most one row.
+// QueryRow always returns a non-nil value. Errors are deferred until
+// Row's Scan method is called.
```

Comment:
// QueryRow executes a query that is expected to return at most one row.
// QueryRow always returns a non-nil value. Errors are deferred until Row's Scan method is called.

Instead of 3 comments, can we have 2 of them added like above ?